nixos/modules/security/pam.nix

```diff
@@ -369,233 +409,272 @@ let
+        }) ++ (optional config.krb5.enable {
+          control = "optional"; path = "${pam_krb5}/lib/security/pam_krb5.so";
+        }) ++ (optional cfg.otpwAuth {
+          control = "optional"; path = "${pkgs.otpw}/lib/security/pam_otpw.so";
+        }) ++ (optional cfg.startSession {
+          control = "optional"; path = "${pkgs.systemd}/lib/security/pam_systemd.so";
+        }) ++ (optional cfg.forwardXAuth {
+          control = "optional"; path = "pam_xauth.so"; arguments = ["xauthpath=${pkgs.xorg.xauth}/bin/xauth" "systemuser=99"];
+        }) ++ (optional (cfg.limits != []) {
+          control = "required"; path = "${pkgs.pam}/lib/security/pam_limits.so"; arguments = ["conf=${makeLimitsConf cfg.limits}"];
+        }) ++ (optional (cfg.showMotd && config.users.motd != null) {
+          control = "optional"; path = "${pkgs.pam}/lib/security/pam_motd.so"; arguments = ["motd=${motd}"];
+        }) ++ (optional (cfg.enableAppArmor && config.security.apparmor.enable) {
+          control = "optional"; path = "${pkgs.apparmor-pam}/lib/security/pam_apparmor.so"; arguments = ["order=user,group,default" "debug"];
+        }) ++ (optional (cfg.enableKwallet) ({
+          control = "optional"; path = "${pkgs.plasma5.kwallet-pam}/lib/security/pam_kwallet5.so";
+          arguments = "kwalletd=${pkgs.kdeFrameworks.kwallet.bin}/bin/kwalletd5";
+        })) ++ (optional (cfg.enableGnomeKeyring) {
+          control = "optional"; path = "${pkgs.gnome3.gnome-keyring}/lib/security/pam_gnome_keyring.so"; arguments = ["auto_start"];
+        }) ++ (optional cfg.gnupg.enable {
+          control = "optional"; path = "${pkgs.pam_gnupg}/lib/security/pam_gnupg.so";
+          arguments = optional cfg.gnupg.noAutostart "no-autostart";
+        }) ++ (optional (config.virtualisation.lxc.lxcfs.enable) {
+          control = "optional"; path = "${pkgs.lxc}/lib/security/pam_cgfs.so"; arguments = [ "-c all" ];
+        });
+      };
     };
-
   };
 
-
   inherit (pkgs) pam_krb5 pam_ccreds;
 
   use_ldap = (config.users.ldap.enable && config.users.ldap.loginPam);
   pam_ldap = if config.users.ldap.daemon.enable then pkgs.nss_pam_ldapd else pkgs.pam_ldap;
 
   # Create a limits.conf(5) file.
-  makeLimitsConf = limits:
-    pkgs.writeText "limits.conf"
-       (concatMapStrings ({ domain, type, item, value }:
-         "${domain} ${type} ${item} ${toString value}\n")
-         limits);
+  makeLimitsConf = limits: pkgs.writeText "limits.conf" (
+    concatMapStrings ({ domain, type, item, value }:
+      "${domain} ${type} ${item} ${toString value}\n"
+    ) limits
+  );
 
   motd = pkgs.writeText "motd" config.users.motd;
 
-  makePAMService = name: service:
-    { name = "pam.d/${name}";
-      value.source = pkgs.writeText "${name}.pam" service.text;
-    };
-
-in
-
-{
+  makePAMService = name: service: let
+    argumentToString = value: if hasInfix " " value then "[${escape [ "]" ] value}]" else value;
```

Comment:
Why do we need the escape here?

Reply:
Nevermind, I RTFMed

> Note, if you wish to include spaces in an argument, you should surround that argument with square brackets.

> When using this convention, you can include `[' characters inside the string, and if you wish to include a `]' character inside the string that will survive the argument parsing, you should use `\]'. In other words: [..[..\]..]    -->   ..[..]..

Reply:
Maybe it should be up to the user to escape any `]` in a single argument that contains spaces. I'm guessing we will miss something here and it might break stuff for people. Let's not do backend stuff like this, users getting to this level of configuration already know about this or will find out from the manpage like I did.

Reply:
I soooo wish I could think of a way to make this more Nixy than a list of strings. For example, if I want to override the umask in the `mkHomeDir` module, I'd have to override the whole list, and I wouldn't be able to do something like `mkHomeDir.args.umask = "0522";` (random value, use at your own risk). Anyone else has a take on this?

Reply:
> Let's not do backend stuff like this, users getting to this level of configuration already know about this or will find out from the manpage like I did.

I disagree pretty strongly with this take, because accidentally adding an argument with spaces doesn't result in an error, it results in a silently different configuration (with that argument treated as multiple separate arguments). You clearly have been configuring PAM modules, and weren't aware of the need to escape spaces. If you had added an argument that contained spaces, it would have silently been accepted but then treated as multiple separate arguments, which could have resulted in an insecure configuration.

I'm also unable to think of a situation in which case us doing the escaping would cause trouble, apart from perhaps double-escaping. But I think people are more likely to read the nix option docs, where we can note that we do this.

Reply:
> I soooo wish I could think of a way to make this more Nixy than a list of strings.

I agree here, but there are lots of options that are just a single string... :shrug: It'd be nice if they had to be key-value pairs.

Reply:
Ah, I have a thought. We can use `types.attrsOf (types.either types.str types.bool)`. Then, if the value is `bool`, we add the key if it is `true`, omit the key if it is `false`. If the value is `str`, we use `key=value` syntax. Could this work?

Reply:
I can't see how that would work. Could you give an example?
`attrsOf (nullOr str)` might be better. It would look like this, assuming we have a `render` function outputting the result from a single argument (name, value pair):
```
>>> render { name = "debug"; value = null; }
"debug"
>>> render { name = "umask"; value = "0027"; }
"umask=0027"
```

Usage:
```
args = {
  debug = null;
  umask = "0027";
};
```

Reply:
I missed this comment, I was thinking as follows:

```
args = {
  debug = true;
  umask = "0027";
}
```

which would render as `"debug umask=0027"`. The `render` function would check whether the attr value is a string, in which case i renders as `"key=value"` or a boolean, in which case it renders `"key"` if the value is `true`, otherwise `""`. The advantage of this is it allows removing a key, by doing `{ args.debug = mkForce false; }`, I'm not sure theres a nice way of unsetting a key from an attrset?